README.md
# Ruby API client for ONTRAPORT

[![Gem Version](https://badge.fury.io/rb/ontraport.svg)](https://badge.fury.io/rb/ontraport)
[![GitHub](https://img.shields.io/badge/github-ontraport--ruby-blue.svg)][repo]
[![Documentation](http://img.shields.io/badge/docs-rdoc.info-blue.svg)][rubydoc]
[![License](http://img.shields.io/badge/license-MIT-yellowgreen.svg)](#license)

Overview
--------

This Gem implements ONTRAPORT's JSON REST API - that's a lot of capitals... but it's how they stylize their name :)

So far, all the [objects API][ontra-objects-api] methods are supported. Forthcoming releases will add support for the other APIs.

Full documentation available [here][rubydoc].

Usage
-----

```ruby
Ontraport.get_object :contact, 12345
#=> #<Ontraport::Response @data=...>
```

```ruby
Ontraport.save_or_update :contact, { email: 'foo@bar.com', firstname: 'Foo' }
#=> #<Ontraport::Response @data=...>
```

```ruby
Ontraport.tag_objects :contact, { add_list: '11111,22222', ids: '33333,44444' }
#=> #<Ontraport::Response @data=...>
```

See [documentation][rubydoc] for more details/examples.

Installation
------------

### Bundler

Add the Ontraport gem to your Gemfile:

```ruby
gem 'ontraport', '~> 0.1'
```

### Manual

Install the Gem from your terminal -

```bash
gem install ontraport
```

Configuration
-------------

In Rails, use an initializer to configure -

```ruby
# config/initializers/ontraport.rb
require 'ontraport'

Ontraport.configure do |config|
  config.api_id = 'foo'
  config.api_key = 'bar'
end
```

Contact
-------

If you have a suggestion or a bug report, please [file an issue][issues].

License
-------

The project uses the MIT License. See LICENSE.md for details.

[issues]: https://github.com/hamzatayeb/ontraport-ruby/issues
[rubydoc]: http://rubydoc.info/gems/ontraport
[repo]: http://github.com/hamzatayeb/ontraport-ruby
[ontra-objects-api]: https://api.ontraport.com/doc/#!/objects/


CHANGELOG.md
# Unreleased

# 0.1.21 - September 6, 2017

- require HTTParty in main module (fixes #1)

# 0.1.2 - April 14, 2016

- Add support for Objects `create` method
- Better formatting for `APIError` exceptions
- Add Transactions `get_order` method

# 0.1.1 - February 6, 2016

- Update documentation

# 0.1 - February 6, 2016

- Initial release
- Implements all "objects" API methods from ONTRAPORT


lib/ontraport/response.rb
module Ontraport
  class Response
    # @!attribute [r] code
    #   Response code
    #
    #   @return [String]
    #
    # @!attribute [r] data
    #   A collection containing actual object data returned from the API.
    #
    #   @return [Array<Hash{String=>String,NilClass}>, Hash{String=>String,NilClass}] either a
    #     Hash containing keys for each field belonging to the object type, or an array of those
    #
    # @!attribute [r] updates
    #   @return [Array]
    #
    # @!attribute [r] notifications
    #   @return [Array]
    #
    # @!attribute [r] account_id
    #   Account Id
    #
    #   @return [String] the Id of your ONTRAPORT account
    #
    # @!attribute [r] misc
    #
    # @!attribute [r] name
    #   Name of an object in the schema.
    #
    #   @return [String] name of the object type, used by the +/objects/meta+ call
    #
    # @!attribute [r] fields
    #   Describes the field metadata for an object.
    #
    #   @return [Hash{String=>Hash}] a collection where keys correspond to field names, and
    #     values are a nested +Hash+ containing the following:
    #     - +'alias'+ - Human-readable name of the field
    #     - +'type'+ - type of field. May be one of the following:
    #       - +check+
    #       - +country+
    #       - +fulldate+
    #       - +list+
    #       - +longtext+
    #       - +numeric+
    #       - +price+
    #       - +phone+
    #       - +state+
    #       - +drop+
    #       - +text+
    #       - +timestamp+
    #       - +parent+
    #       - +subscription+
    #       - +email+
    #       - +sms+
    #     - +'options'+ - either an Array or a Hash of options. This is +nil+ unless +type+ is +list+ or +drop+
    #
    # @!attribute [r] schema_object_id
    #   Numeric object identifier for a particular object in the schema.
    #
    #   @return [String]
    attr_reader :code, :data, :updates, :notifications, :account_id, :misc, :name, :fields, :schema_object_id

    def initialize args
      args.each do |key,val|
        instance_variable_set("@#{key}", val) unless val.nil?
      end
    end
  end
end

lib/ontraport.rb
require "httparty"

Dir[File.expand_path('../ontraport/*.rb', __FILE__)].each do |file|
  require file
end

# @author Hamza Tayeb
# @see https://api.ontraport.com/doc ONTRAPORT API Documentation
module Ontraport
  BASE_URL = 'https://api.ontraport.com/'
  API_VERSION = '1'

  # Describe a given object type, including the numeric object Id and the available fields.
  #
  # @example
  #   Ontraport.describe :contact
  #
  # @param object [Symbol] the type of object you want to describe
  # @raise [ArgumentError] if the argument is not a +Symbol+
  # @raise [ObjectNotFoundError] if the argument does not match any object defined in the schema
  # @return [Hash{String=>String,Hash}] object metadata. Keys are:
  #   - +'name'+ - name of the object
  #   - +'fields'+ - Hash containing the object's fields
  #   - +'schema_object_id'+ - numeric Id of the object, as a String
  def self.describe object
    unless object.class.eql? Symbol
      raise ArgumentError.new "Must provide a symbol for the object name."
    end

    unless metadata = objects_meta.data.find{ |k, v| v['name'].downcase.to_sym.eql? object }
      raise ObjectNotFoundError.new "No object matching #{object.inspect} could be found."
    end

    metadata.second.update 'schema_object_id' => metadata.first
  end

  # Clear the cache of object metadata generated by the describe call. Use this
  # if you make changes to custom fields or objects in your ONTRAPORT instance and
  # don't want to reload the Ontraport module.
  #
  # @example
  #   Ontraport.clear_describe_cache!
  #
  # @return [nil]
  def self.clear_describe_cache!
    @objects_meta_cache = nil
  end

  # @!group "Objects" Methods

  # Retrieve a single object of the specified type.
  #
  # @example
  #   Ontraport.get_object :contact, 12345
  #   #=> #<Ontraport::Response @data=...>
  #
  # @param object_type [Symbol] the type of object, for instance +:contact+
  # @param id [Integer] Id of the record you want
  # @return [Response]
  def self.get_object object_type, id
    objects_call :get, object_type, endpoint: '/object', data: { id: id }
  end

  # Retrieve a collection of objects of the specified type, matching the query
  # specified by the parameters.
  #
  # @example
  #   Ontraport.get_objects :contact, { condition: "email like '%@foo.com'", sort: 'lastname' }
  #
  # @see https://api.ontraport.com/doc/#!/objects/getObjects List of accepted params
  #
  # @param object_type [Symbol] the type of object
  # @param params [Hash] hash containing request parameters
  # @return [Response]
  def self.get_objects object_type, params={}
    objects_call :get, object_type, endpoint: '/objects', data: params
  end

  # Create an object with the given data
  #
  # @example
  #   Ontraport.create :contact, { email: 'foo@bar.com', firstname: 'Foo' }
  #   #=> #<Ontraport::Response @data=...>
  #
  # @see https://api.ontraport.com/doc/#!/objects/createObject API docs
  #
  # @param object_type [Symbol] the type of object
  # @param params [Hash] input data
  # @return [Response]
  def self.create object_type, params
    objects_call :post, object_type, endpoint: '/objects', data: params
  end

  # Create an object with the given data, or merge if the unique field matches another row.
  #
  # @example
  #   Ontraport.save_or_update :contact, { email: 'foo@bar.com', firstname: 'Foo' }
  #   #=> #<Ontraport::Response @data=...>
  #
  # @see https://api.ontraport.com/doc/#!/objects/createormergeObject API docs
  #
  # @param object_type [Symbol] the type of object
  # @param params [Hash] input data
  # @return [Response]
  def self.save_or_update object_type, params
    objects_call :post, object_type, endpoint: '/objects/saveorupdate', data: params
  end

  # Given an object type, Id, and a +Hash+ of changes, update an single row.
  #
  # @example
  #   Ontraport.update_object :contact, 12345, { firstname: 'ChangeMe' }
  #   #=> #<Ontraport::Response @data=...>
  #
  # @param object_type [Symbol] the type of object
  # @param params [Hash] update data. Use +.describe+ to get a list of available fields.
  # @return [Response]
  def self.update_object object_type, id, params
    objects_call :put, object_type, endpoint: '/objects', data: params.update(id: id)
  end

  # Add tags to objects matching the supplied conditions. In addition to various conditionals, you
  # may supply a list of Ids for objects you wish to tag (+ids+). The +add_list+ parameter (which contains the
  # Ids of the tags you want to add) is required.
  #
  # @note The +add_list+ and +ids+ parameters should be strings comprised of comma-delimeted Integers.
  #
  # @example
  #   Ontraport.tag_objects :contact, { add_list: '11111,22222', ids: '33333,44444' }
  #   #=> #<Ontraport::Response @data=...>
  #
  # @see https://api.ontraport.com/doc/#!/objects/addTag API docs
  #
  # @param object_type [Symbol] the type of object
  # @param params [Hash] parameters describing the conditions of the tag operation.
  # @return [Response]
  def self.tag_objects object_type, params

    objects_call :put, object_type, endpoint: '/objects/tag', data: params
  end

  # Remove tags from objects matching the supplied conditions. Interface is nearly identical to +#tag_objects+
  #
  # @note This method expects +remove_list+ as a required parameter.
  #
  # @see tag_objects
  # @see https://api.ontraport.com/doc/#!/objects/removeTag API docs
  #
  # @param object_type [Symbol] the type of object
  # @param params [Hash] parameters describing the conditions of the tag operation.
  # @return [Response]
  def self.untag_objects object_type, object:, tag:
    objects_call :delete, object_type, endpoint: '/objects/tag', data: params
  end

  # @!endgroup
  # @!group "Transactions" Methods

  # Get full information about an order
  #
  # @see https://api.ontraport.com/doc/#!/transactions/getOrder API docs
  #
  # @param order_id [Integer] Id of the order
  # @return [Response]
  def self.get_order order_id
    request_with_authentication :get, endpoint: '/transaction/order', data: { id: order_id }
  end

  # @!endgroup

  private
    def self.request_with_authentication method, endpoint:, data: nil
      data_param = method.eql?(:get) ? :query : :body

      args = [method, "#{BASE_URL}#{API_VERSION}#{endpoint}"]
      kwargs = {
        :headers => { 'Api-Appid' => @configuration.api_id, 'Api-Key' => @configuration.api_key },
        data_param => data
      }

      response = HTTParty.send *args, **kwargs

      unless response.code.eql? 200
        error = "#{response.code} #{response.msg}"
        raise APIError.new(response.body.present? ? "#{error} - #{response.body}" : error)
      end

      Response.new **response.parsed_response.symbolize_keys
    end

    def self.objects_call method, object_type, endpoint:, data: {}
      metadata = describe object_type
      data.update 'objectID' => metadata['schema_object_id']

      request_with_authentication method, endpoint: endpoint, data: data
    end

    def self.objects_meta
      @objects_meta_cache ||= request_with_authentication :get, endpoint: '/objects/meta'
    end
end
